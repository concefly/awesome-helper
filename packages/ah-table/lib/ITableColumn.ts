export interface ITableColumn {
  dataIndex: string;
  title?: string;
  width?: number;
}
