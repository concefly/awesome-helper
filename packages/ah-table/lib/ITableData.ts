export interface ITableData {
  [name: string]: any;
}
