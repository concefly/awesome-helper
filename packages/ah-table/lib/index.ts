export * from './ITableColumn';
export * from './ITableData';
export * from './ITableImplement';
export * from './Table';
export * from './TableBody';
export * from './TableCell';
export * from './TableHead';
export * from './TableTd';
export * from './TableTh';
export * from './TableTr';
