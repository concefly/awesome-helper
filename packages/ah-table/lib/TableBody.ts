import { ITableImplement } from './ITableImplement';
import { TableTr } from './TableTr';

export class TableBody {
  constructor(readonly imp: ITableImplement, readonly trList: TableTr[]) {}

  public render() {
    return this.trList.map(t => t.render()).join('\n');
  }
}
