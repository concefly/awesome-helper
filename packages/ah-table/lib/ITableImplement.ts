import { TableBody } from './TableBody';
import { TableHead } from './TableHead';
import { TableTd } from './TableTd';
import { TableTh } from './TableTh';
import { TableTr } from './TableTr';

export type ITableImplement = {
  TableHead: typeof TableHead;
  TableBody: typeof TableBody;
  TableTd: typeof TableTd;
  TableTh: typeof TableTh;
  TableTr: typeof TableTr;
};
