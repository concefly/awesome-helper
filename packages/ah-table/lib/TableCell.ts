import * as _ from 'lodash';
import { ITableImplement } from './ITableImplement';

export class TableCell {
  constructor(
    readonly imp: ITableImplement,
    readonly data: any,
    readonly config: {
      width: number;
    }
  ) {}

  public render() {
    const strWidth = this.config.width - 2;
    let str =
      typeof this.data === 'string'
        ? this.data
        : typeof this.data === 'number'
        ? this.data + ''
        : JSON.stringify(this.data) || '-';

    str = str.length > strWidth ? _.truncate(str, { length: strWidth }) : _.padEnd(str, strWidth);

    str = ` ${str} `;

    return str;
  }
}
