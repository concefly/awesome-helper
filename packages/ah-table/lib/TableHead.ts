import { ITableImplement } from './ITableImplement';
import { TableTr } from './TableTr';

export class TableHead {
  constructor(readonly ioc: ITableImplement, readonly tr: TableTr) {}

  public render() {
    return this.tr.render();
  }
}
