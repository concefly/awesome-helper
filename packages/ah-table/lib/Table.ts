import { ITableColumn } from './ITableColumn';
import { ITableData } from './ITableData';
import { ITableImplement } from './ITableImplement';
import { TableBody } from './TableBody';
import { TableHead } from './TableHead';
import { TableTd } from './TableTd';
import { TableTh } from './TableTh';
import { TableTr } from './TableTr';
import * as _ from 'lodash';

export class Table<T extends ITableData = ITableData> {
  protected imp: ITableImplement = {
    TableHead,
    TableBody,
    TableTd,
    TableTh,
    TableTr,
  };

  protected defaultColumnWidth = 16;

  constructor(readonly columns: ITableColumn[], readonly datasource: T[]) {}

  public getWidth() {
    return _.sum(this.columns.map(c => c.width || this.defaultColumnWidth));
  }

  public render() {
    const headTr = new this.imp.TableTr(
      this.imp,
      this.columns.map(
        c =>
          new this.imp.TableTh(this.imp, c.title || c.dataIndex, {
            width: c.width || this.defaultColumnWidth,
          })
      )
    );
    const head = new this.imp.TableHead(this.imp, headTr);

    const bodyTrList = this.datasource.map(data => {
      return new this.imp.TableTr(
        this.imp,
        this.columns.map(
          c =>
            new this.imp.TableTd(this.imp, _.get(data, c.dataIndex), {
              width: c.width || this.defaultColumnWidth,
            })
        )
      );
    });

    const body = new this.imp.TableBody(this.imp, bodyTrList);
    const gutter = Array(this.getWidth()).fill('-').join('');

    return ['', gutter, head.render(), gutter, body.render(), gutter, ''].join('\n');
  }
}
