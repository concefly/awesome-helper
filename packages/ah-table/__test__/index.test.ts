import { ITableColumn } from '../lib/ITableColumn';
import { ITableData } from '../lib/ITableData';
import { Table } from '../lib/Table';

const columns: ITableColumn[] = [
  {
    dataIndex: 'id',
    width: 5,
  },
  {
    title: 'fruit title',
    dataIndex: 'fruit',
    width: 30,
  },
  {
    dataIndex: 'date',
  },
  {
    dataIndex: 'foo.bar',
  },
];

const datasource: ITableData[] = [
  { fruit: 'apple', date: '2020-11-11' },
  { fruit: 'banana', foo: { bar: 'xxx' } },
  { fruit: 'long name wwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwww' },
].map((d, id) => ({ id, ...d }));

describe('Table', () => {
  it('normal', () => {
    const t = new Table(columns, datasource);
    const s = t.render();
    expect(s).toMatchSnapshot();
  });
});
