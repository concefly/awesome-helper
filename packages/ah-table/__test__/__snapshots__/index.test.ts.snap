// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table normal 1`] = `
"
---------------------------------------------------
 id  | fruit title                  | date           
---------------------------------------------------
 0   | apple                        | 2020-11-11     
 1   | banana                       | -              
 2   | long name wwwwwwwwwwwwwww... | -              
---------------------------------------------------
"
`;
